asset/file_reader/index.js

```diff
@@ -0,0 +1,155 @@
+    const opConfig = executionContext.config.operations[0];
+    const events = context.foundation.getEventEmitter();
+    checkProvidedPath(opConfig.path);
+    const queue = new Queue();
+
+    // Set up flags for slice generation/management
+    let isFinished = false;
+    let isShutdown = false;
+
+    events.on('worker:shutdown', () => {
+        isShutdown = true;
+    });
+
+    function checkProvidedPath(filePath) {
+        try {
+            const dirStats = fse.lstatSync(filePath);
+            if (dirStats.isSymbolicLink()) {
+                throw new Error('Directory for `file_reader` cannot be a symlink!');
+            }
+            const dirContents = fse.readdirSync(filePath);
+            if (dirContents.length === 0) {
+                throw new Error('Must provide a non-empty directory for `file_reader`!!');
+            }
+        } catch (err) {
+            throw new Error(`Path must be valid! Encountered the following error:\n${err}`);
+        }
+    }
+
+    async function processFile(filePath) {
+        const stat = await fse.stat(filePath);
+        const total = stat.size;
+        getOffsets(opConfig.size, total, opConfig.delimiter).forEach((offset) => {
+            offset.path = filePath;
+            queue.enqueue(offset);
+        });
+        return true;
+    }
+
+    async function getFilePaths(filePath) {
+        const dirContents = await fse.readdir(filePath);
+        return Promise.map(dirContents, async (file) => {
+            const fullPath = path.join(filePath, file);
+            const stats = await fse.lstat(fullPath);
+            if (stats.isFile()) {
+                return processFile(fullPath);
+            }
+            if (stats.isDirectory()) {
+                return getFilePaths(fullPath);
+            }
+            logger.error(`${file} is not a file or directory!!`);
+            return true;
+        })
+            // Catch the error and log it so the execution controller doesn't crash and burn if
+            // there is a bad file or directory
+            .catch((err) => {
+                const error = new TSError(err, {
+                    reason: 'Error while reading file or directory',
+                    context: {
+                        filePath
+                    }
+                });
+                logger.error(error);
+            });
+    }
+
+    getFilePaths(opConfig.path)
+        .then(() => {
+            isFinished = true;
+        });
+
+    return [() => {
+        // Grab a record if there is one ready in the queue
+        if (queue._size > 0) return queue.dequeue();
+
+        // Finish slicer if the queue is empty and it's done prepping slices
+        if (isFinished) return null;
+
+        // If the queue is empty, wait for a new slice or throw an error if the slicer shuts down
+        // early
+        return new Promise((resolve, reject) => {
+            const intervalId = setInterval(() => {
+                if (queue._size > 0) {
```

Comment:
It is best to use the public method to get the size instead of a "private" variable.

```suggestion
                if (queue.size() > 0) {
```

Reply:
fixed in both cases